lib/generic-request.js

```diff
@@ -0,0 +1,27 @@
+const isCorrectHttpMethod = require('./check-http-method');
+const HTTP_STATUS_CODES = require('./http-status-codes');
+const rxjs = require('rxjs');
+const Observable = rxjs.Observable;
+const throwError = rxjs.throwError;
+const request = require('request');
+const isFailedResponse = statusCode => statusCode >= HTTP_STATUS_CODES.BAD_REQUEST;
+
+module.exports = (method, url, options) => {
+    if (!isCorrectHttpMethod(method)) {
+        const errMsg = `httpRx client was given an invalid method name for http request: ${method}`;
+        return throwError(new Error(errMsg));
+    }
+    return new Observable(observer => {
+        request[method](url, options, (err, response, body) => {
+            const responseWrapper = {response, body};
+            if (err) {
+                return observer.error(err);
+            }
+            if (isFailedResponse(response.statusCode)) {
+                return observer.error(responseWrapper);
+            }
+            observer.next(responseWrapper);
+            observer.complete(responseWrapper);
+        });
+    });
```

Comment:
If I remember that correctly, you may return a method that will provide the client ability to cancel observable during invocation. Could we provide here that logic to support back-pressure?

Reply:
Sounds interesting.
@blackorion, if taken look at the [example](https://www.learnrxjs.io/operators/creation/create.html) you provided, but still I see only glimpse of the original idea.
Can you, please, clarify it a bit further?

Reply:
Sure. The main idea behind the callback that can cancel the request is to provide the ability for the procedure chain to override previous request. for example let's say you're updating field in the form and user does update these fields quicker than the request is processed by BE. In that case you probably would want to cancel the previous request and send the new one. (we're ignoring the fact you're can add denounce methods in the chain and other tricks, that are by the way based on the same idea)

Reply:
I've created a small (preview)[https://rxviz.com/v/2ORyZ04o] for that, hope will be helpful. Checkout the diff between the console and what you're see in the visualization. Try it with and without return callback

Reply:
Thanks, I'll take a look at it.

Reply:
An [official documentation for Observable.create](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-create) also sheds some light on what is a teardown logic and what it is used for.
I've added request cancellation, when it comes to unsubscribing from an observable.
Request package has a convenient method called `abort`, which one might want to call on a `request.Request` instance in order to cancel pending request.

Reply:
I'll add a respective unit test later.
It has to check, that a pending request is really cancelled once we decided to unsubscribe from an observable.